### PR TITLE
throw out error if index hint not exist

### LINKED ISSUE #4792: Check the index validity when executing `use index ...`
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
create table tt(a int);
select count(*) from tt use index (idx);
```

2. What did you expect to see?
```
mysql>  select count(*) from tt use index (idx);
ERROR 1176 (42000): Key 'idx' doesn't exist in table 'tt'
```

3. What did you see instead?
```
tidb> select count(*) from tt use index (idx);
+----------+
| count(*) |
+----------+
|        0 |
+----------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by ceshihao**
Try to fix #4792

### CODE REVIEW

File: plan/preprocess.go [hunk 1]
```diff
@@ -561,3 +563,22 @@ func (p *preprocessor) resolveAlterTableStmt(node *ast.AlterTableStmt) {
 		}
 	}
 }
+
+func (p *preprocessor) checkSelectStmt(node *ast.SelectStmt) {
```
Comment (shenli) [inline plan/preprocess.go]: Please add comments.

**Comment by winoros** [comment]
please fix CI and add unit test.

**Comment by hanfei1991** [comment]
you should add check in `availableIndices`

**Comment by ceshihao** [comment]
@hanfei1991 thx, i will take a look at it

**Comment by winoros** [comment]
Any update? @ceshihao

**Comment by ceshihao** [comment]
update `availableIndices` for error

File: plan/physical_plan_builder.go [hunk 2]
```diff
@@ -180,7 +180,7 @@ func (p *LogicalJoin) getIndexJoinByOuterIdx(outerIdx int) []PhysicalPlan {
 	if !ok || x.unionScanSchema != nil {
 		return nil
 	}
-	indices, includeTableScan := availableIndices(x.indexHints, x.tableInfo)
+	indices, includeTableScan, _ := availableIndices(x.indexHints, x.tableInfo)
```
Comment (hanfei1991) [inline plan/physical_plan_builder.go]: You should check error here but not in logical_plan_builder

**Comment by winoros** [comment]
Please `make dev` first in your local environment to make the CI pass.

File: plan/physical_plan_builder.go (hunk 2 above)
Comment (ceshihao) [reply plan/physical_plan_builder.go]: move hint index missing error check from logical_plan_builder to physical_plan_builder

**Comment by ceshihao** [comment]
move hint index missing error check from logical_plan_builder to physical_plan_builder.

I have to add an error return for `generatePhysicalPlans`, so that error can be passed back.

**Comment by hanfei1991** [comment]
Please extract all indices once in logical plan builder and hold them. So we needn't call `availableIndices` every time.

**Comment by ceshihao** [comment]
resolve the conflict

**Comment by winoros** [comment]
@ceshihao You can store the available indices as a field of `DataSource`, to make code clean and avoid some error check.

**Comment by ceshihao** [comment]
store `availableIndices` in `DataSource`, and resolve conflict

File: mysql/errcode.go [hunk 3]
```diff
@@ -882,6 +882,7 @@ const (
 	ErrMustChangePasswordLogin                                      = 1862
 	ErrRowInWrongPartition                                          = 1863
 	ErrErrorLast                                                    = 1863
+	ErrHintMissIndex                                                = 1864
```
[the remark this answers is not in the document]
Comment (ceshihao) [reply mysql/errcode.go]: use ErrKeyDoesNotExits 1176 in 10aff85

File: plan/dag_plan_test.go [hunk 4]
```diff
@@ -190,6 +185,40 @@ func (s *testPlanSuite) TestDAGPlanBuilderSimpleCase(c *C) {
 	}
 }
 
+func (s *testPlanSuite) TestDAGPlanBuilderOptimizeError(c *C) {
```
[the remark this answers is not in the document]
Comment (ceshihao) [reply plan/dag_plan_test.go]: add ut in 1a2bc3e

File: plan/logical_plan_test.go [hunk 5]
```diff
@@ -660,6 +660,28 @@ func (s *testPlanSuite) TestPlanBuilder(c *C) {
 	}
 }
 
+func (s *testPlanSuite) TestPlanBuilderError(c *C) {
```
Comment (hanfei1991) [inline plan/logical_plan_test.go]: We have a test called `TestValidate`. I'm sorry I made a mistake in last comment.

Comment (ceshihao) [reply plan/logical_plan_test.go]: Move ErrKeyDoesNotExits check ut to `TestValidate` in 98d084b

Sorry, I am not very familiar with the case organization.

File: plan/planbuilder.go [hunk 6]
```diff
@@ -43,6 +43,7 @@ var (
 	ErrBadGeneratedColumn   = terror.ClassOptimizerPlan.New(CodeBadGeneratedColumn, mysql.MySQLErrName[mysql.ErrBadGeneratedColumn])
 	ErrFieldNotInGroupBy    = terror.ClassOptimizerPlan.New(CodeFieldNotInGroupBy, mysql.MySQLErrName[mysql.ErrFieldNotInGroupBy])
 	ErrBadTable             = terror.ClassOptimizerPlan.New(CodeBadTable, mysql.MySQLErrName[mysql.ErrBadTable])
+	ErrKeyDoesNotExits      = terror.ClassOptimizerPlan.New(CodeKeyDoesNotExits, mysql.MySQLErrName[mysql.ErrKeyDoesNotExits])
```
Comment (hanfei1991) [inline plan/planbuilder.go]: Exit -> Exist

Comment (ceshihao) [reply plan/planbuilder.go]: fixed typo

**Comment by winoros** [comment]
@hanfei1991  PTAL

**Pull request merged**